### PR TITLE
Support parsing SQL Server INSERT INTO sql #29152

### LINKED ISSUE #29152: Support parsing SQL Server INSERT INTO sql
# Background
Hi community.
This issue is for #29149.

The ShardingSphere SQL parser engine helps users to parse SQL to create the AST (Abstract Syntax Tree) and visit the AST to get SQLStatement (Java Object). Currently, we are planning to enhance the support for `SQL Server` SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

# Task
This issue is to support more `SQL Server` sql parsing, as follows:
```sql
INSERT INTO dbo.T1 VALUES (1, N'Natalia')
```
[link](https://learn.microsoft.com/zh-cn/sql/tools/bcp-utility?view=sql-server-ver16)

```sql
INSERT INTO dbo.T1 VALUES (2, N'Mark')
```
[link](https://learn.microsoft.com/zh-cn/sql/tools/bcp-utility?view=sql-server-ver16)

```sql
INSERT INTO dbo.T1 VALUES (3, N'Randolph')
```
[link](https://learn.microsoft.com/zh-cn/sql/tools/bcp-utility?view=sql-server-ver16)

```sql
SELECT x.$(ColumnName)
FROM Person.Person x
WHERE x.BusinessEntityID < 5
```
[link](https://learn.microsoft.com/zh-cn/sql/tools/sqlcmd/sqlcmd-use-scripting-variables?view=sql-server-ver16)

```sql
INSERT INTO AdventureWorks2022.dbo.VariableTest(Col1)
VALUES('$(tablename)')
```
[link](https://learn.microsoft.com/zh-cn/sql/tools/sqlcmd/sqlcmd-use-scripting-variables?view=sql-server-ver16)

# Process
1. First confirm that this is a correct `SQL Server` sql syntax, if not please leave a message under the issue and ignore it;
2. Compare SQL definitions in Official SQL Doc and ShardingSphere SQL Doc;
3. If there is any difference in ShardingSphere SQL Doc, please correct them by referring to the Official SQL Doc;
4. Run mvn install the current_file_module;
5. Check whether there are any exceptions. If indeed, please fix them. (Especially xxxVisitor.class);
6. Add new corresponding SQL case in SQL Cases and expected parsed result in Expected Statement XML;
7. Run SQLParserParameterizedTest to make sure no exceptions.

# Relevant Skills
1. Master JAVA language
2. Have a basic understanding of Antlr `g4` file
3. Be familiar with `SQL Server` SQLs

### PR DESCRIPTION
**Pull request by TherChenYang**
Fixes #29152.

Changes proposed in this pull request:
- Support parsing SQL Server INSERT INTO sql
1. INSERT INTO dbo.T1 VALUES (1, N'Natalia')
2. SELECT x.$(ColumnName) FROM Person.Person x WHERE x.BusinessEntityID < 5
3. INSERT INTO AdventureWorks2022.dbo.VariableTest(Col1) VALUES('$(tablename)')

- Fix the error parser in oracle Nchar
---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes

### CODE REVIEW

**Review by zihaoAK47 — CHANGES_REQUESTED** [review]

**Comment by strongduanmu** [comment]
Hi @TherChenYang, can you refer the original issue? I will assign it to you.

File: parser/sql/dialect/sqlserver/src/main/antlr4/imports/sqlserver/BaseRule.g4 [hunk 1]
```diff
@@ -151,7 +152,7 @@ sequenceName
     ;
 
 tableName
-    : (owner DOT_)? name
+    : ((databaseName DOT_)? owner DOT_)? name
```
[the remark this answers is not in the document]
Comment (zihaoAK47) [reply parser/sql/dialect/sqlserver/src/main/antlr4/imports/sqlserver/BaseRule.g4:155]: We can use the selection mode of antlr.

File: parser/sql/dialect/sqlserver/src/main/antlr4/imports/sqlserver/BaseRule.g4 [hunk 2]
```diff
@@ -167,7 +168,11 @@ serviceName
     ;
 
 columnName
-    : (owner DOT_)? name
+    : (owner DOT_)? (name | scriptVariableName)
+    ;
+
+scriptVariableName
+    : DOLLAR_ LP_ name RP_
```
[the remark this answers is not in the document]
Comment (zihaoAK47) [reply parser/sql/dialect/sqlserver/src/main/antlr4/imports/sqlserver/BaseRule.g4:175]: Okay, I understand. Thank you for your explanation
ref: https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-use-scripting-variables?view=sql-server-ver16

**Review by zihaoAK47 — APPROVED** [review]
Great PR, thank you

**Pull request merged**